File: nodejs/packages/periph/src/chips/gyroscope/l3gd20h.js
'use strict';

const _REG_WHO_AM_I      = 0x0F;
const _REG_CTRL_REG1     = 0x20;
const _REG_CTRL_REG2     = 0x21;
const _REG_CTRL_REG3     = 0x22;
const _REG_CTRL_REG4     = 0x23;
const _REG_CTRL_REG5     = 0x24;
const _REG_OUT_TEMP      = 0x26;
const _REG_STATUS        = 0x27;
const _REG_OUT_X_L       = 0x28;
const _REG_OUT_X_H       = 0x29;
const _REG_OUT_Y_L       = 0x2A;
const _REG_OUT_Y_H       = 0x2B;
const _REG_OUT_Z_L       = 0x2C;
const _REG_OUT_Z_H       = 0x2D;
const _REG_FIFO_CTRL     = 0x2E;
const _REG_FIFO_SRC      = 0x2F;
const _REG_INT1_CFG      = 0x30;
const _REG_INT1_SRC      = 0x31;
const _REG_INT1_TSH_XH   = 0x32;
const _REG_INT1_TSH_XL   = 0x33;
const _REG_INT1_TSH_YH   = 0x34;
const _REG_INT1_TSH_YL   = 0x35;
const _REG_INT1_TSH_ZH   = 0x36;
const _REG_INT1_TSH_ZL   = 0x37;
const _REG_INT1_DURATION = 0x38;

const _WHO_AM_I_L3GD20  = 0xD4;
const _WHO_AM_I_L3GD20H = 0xD7;
const _CTRL_REG1_DEFAULT = 0x0F;
const _CTRL_REG4_DEFAULT = 0x80;

const _SENSITIVITY = {
    250:  8.75e-3,
    500:  17.5e-3,
    2000: 70.0e-3,
};

const _DPS_TO_RAD = Math.PI / 180.0;

function _int16Le(data, offset) {
    let v = data[offset] | (data[offset + 1] << 8);
    if (v & 0x8000) v -= 0x10000;
    return v;
}

class L3GD20HMinimal {
    /**
     * L3GD20H (and L3GD20) three-axis MEMS gyroscope — minimal interface.
     *
     * Provides angular rate readings on the X, Y, and Z axes with no
     * configuration beyond the connection. I²C address is 0x6A (SA0/SDO=GND) or
     * 0x6B (SA0/SDO=VCC). SPI uses Mode 3 (CPOL=CPHA=1) by default.
     *
     * Default configuration (baked in at construction):
     *     - 95 Hz ODR, default bandwidth (DR=00, BW=00)
     *     - ±250 dps full scale (sensitivity 8.75 mdps/digit)
     *     - BDU=1 (block data update — hold registers until MSB+LSB read)
     *     - All axes enabled, normal power mode
     *     - 250 ms startup delay for gyroscope stabilization
     *
     * @param {import('../../connection/connection').Connection} connection - Configured I²C or SPI connection.
     * @param {string} [busType='i2c'] - Bus type: 'i2c' or 'spi'.
     */
    constructor(connection, busType = 'i2c') {
        this._conn = connection;
        this._busType = busType;
        this._fullScale = 250;
        this._init();
    }

    async _init() {
        try {
            const who = (await this._readReg(_REG_WHO_AM_I, 1))[0];
            if (who !== _WHO_AM_I_L3GD20 && who !== _WHO_AM_I_L3GD20H) return;
            await this._writeReg(_REG_CTRL_REG4, _CTRL_REG4_DEFAULT);
            await this._writeReg(_REG_CTRL_REG1, _CTRL_REG1_DEFAULT);
            // 250 ms startup delay
            await new Promise(resolve => setTimeout(resolve, 250));
        } catch (e) { /* bus may be idle */ }
    }

    async _writeReg(reg, value) {
        const addr = this._busType === 'spi' ? (reg & 0x3F) : reg;
        await this._conn.write(Buffer.from([addr, value & 0xFF]));
    }

    async _readReg(reg, n) {
        let addr;
        if (this._busType === 'spi') {
            addr = reg | 0xC0;  // READ=1, MS=1 (auto-increment)
        } else if (n > 1) {
            addr = reg | 0x80;  // MSB set = I²C multi-byte auto-increment
        } else {
            addr = reg;
        }
        return this._conn.writeRead(Buffer.from([addr & 0xFF]), n);
    }

    _sensitivity() {
        return _SENSITIVITY[this._fullScale];
    }

    /**
     * Read angular rate on all three axes as a single burst transaction.
     * @returns {Promise<[number, number, number]>} (x_rad_s, y_rad_s, z_rad_s).
     */
    async gyro() {
        const buf = await this._readReg(_REG_OUT_X_L, 6);
        const sens = this._sensitivity();
        const x_dps = _int16Le(buf, 0) * sens;
        const y_dps = _int16Le(buf, 2) * sens;
        const z_dps = _int16Le(buf, 4) * sens;
        return [x_dps * _DPS_TO_RAD, y_dps * _DPS_TO_RAD, z_dps * _DPS_TO_RAD];
    }
}

class L3GD20HFull extends L3GD20HMinimal {
    /**
     * L3GD20H full interface — extends L3GD20HMinimal with full configuration,
     * FIFO, high-pass filter, interrupts, axis-enable, and power-mode control.
     *
     * Adds ODR/bandwidth/full-scale configuration, FIFO with all five modes
     * and watermark, high-pass filter with selectable cutoff, per-axis
     * interrupt generation with threshold and duration, INT1 pin routing,
     * and access to temperature and status registers.
     *
     * @param {import('../../connection/connection').Connection} connection - Configured I²C or SPI connection.
     * @param {string} [busType='i2c'] - Bus type: 'i2c' or 'spi'.
     */
    constructor(connection, busType = 'i2c') {
        super(connection, busType);
        this._odr = 0;
        this._bw = 0;
        this._thresholdRaw = 0;
    }

    /** Output data rate codes (DR[1:0] in CTRL_REG1). */
    static ODR_95_HZ  = 0;
    static ODR_190_HZ = 1;
    static ODR_380_HZ = 2;
    static ODR_760_HZ = 3;

    /** Full-scale codes (0=±250, 1=±500, 2=±2000 dps). */
    static FS_250_DPS  = 0;
    static FS_500_DPS  = 1;
    static FS_2000_DPS = 2;

    /** FIFO modes (FM[2:0] in FIFO_CTRL_REG). */
    static FIFO_BYPASS             = 0;
    static FIFO_FIFO               = 1;
    static FIFO_STREAM             = 2;
    static FIFO_BYPASS_TO_STREAM   = 3;
    static FIFO_STREAM_TO_FIFO     = 7;

    /** HPF modes (HPM[1:0] in CTRL_REG2). */
    static HPM_NORMAL      = 0;
    static HPM_REFERENCE   = 1;
    static HPM_NORMAL_ALT  = 2;
    static HPM_AUTORESET   = 3;

    /** Power mode strings. */
    static POWER_NORMAL     = 'normal';
    static POWER_SLEEP      = 'sleep';
    static POWER_POWERDOWN  = 'power_down';

    /**
     * Configure ODR, bandwidth, and full scale in one call.
     * @param {number} odr         ODR code 0-3 (95/190/380/760 Hz).
     * @param {number} bw          Bandwidth code 0-3 (ODR-dependent).
     * @param {number} fullScale   Full-scale code 0=±250, 1=±500, 2=±2000 dps.
     */
    async configure(odr, bw, fullScale) {
        if (odr < 0 || odr > 3) throw new Error('odr must be 0..3');
        if (bw < 0 || bw > 3) throw new Error('bw must be 0..3');
        if (fullScale < 0 || fullScale > 2) throw new Error('fullScale must be 0..2');
        this._odr = odr;
        this._bw = bw;
        const fsMap = [250, 500, 2000];
        this._fullScale = fsMap[fullScale];
        const ctrl1 = _CTRL_REG1_DEFAULT | ((this._odr & 0x3) << 6) | ((this._bw & 0x3) << 4);
        await this._writeReg(_REG_CTRL_REG1, ctrl1);
        await this._writeReg(_REG_CTRL_REG4, _CTRL_REG4_DEFAULT | ((fullScale & 0x3) << 4));
    }

    /**
     * Read raw 16-bit signed angular rate values.
     * @returns {Promise<[number, number, number]>} (x_raw, y_raw, z_raw).
     */
    async gyroRaw() {
        const buf = await this._readReg(_REG_OUT_X_L, 6);
        return [_int16Le(buf, 0), _int16Le(buf, 2), _int16Le(buf, 4)];
    }

    /**
     * Read the relative temperature count.
     * OUT_TEMP is an 8-bit signed value with 1 LSB/°C sensitivity. There is
     * no absolute calibration — it represents change from the device's
     * power-on temperature baseline. Do not convert to absolute Celsius.
     * @returns {Promise<number>} Signed 8-bit temperature count.
     */
    async temperature() {
        let raw = (await this._readReg(_REG_OUT_TEMP, 1))[0];
        if (raw & 0x80) raw -= 0x100;
        return raw;
    }

    /**
     * Check whether a new X/Y/Z sample is ready.
     * @returns {Promise<boolean>} True if STATUS_REG.ZYXDA (bit 3) is set.
     */
    async dataReady() {
        const status = (await this._readReg(_REG_STATUS, 1))[0];
        return Boolean(status & 0x08);
    }

    /**
     * Configure the high-pass filter (CTRL_REG2).
     * @param {number} mode   HPF mode 0-3 (HPM[1:0]).
     * @param {number} cutoff HPF cutoff code 0-15 (HPCF[3:0]).
     */
    async configureHpFilter(mode, cutoff) {
        if (mode < 0 || mode > 3) throw new Error('mode must be 0..3');
        if (cutoff < 0 || cutoff > 15) throw new Error('cutoff must be 0..15');
        await this._writeReg(_REG_CTRL_REG2, ((mode & 0x3) << 4) | (cutoff & 0x0F));
    }

    /**
     * Enable or disable the high-pass filter on the output path.
     * @param {boolean} enable True to enable (sets HPen in CTRL_REG5), False to disable.
     */
    async enableHpFilter(enable) {
        const ctrl5 = (await this._readReg(_REG_CTRL_REG5, 1))[0];
        await this._writeReg(_REG_CTRL_REG5, enable ? (ctrl5 | 0x10) : (ctrl5 & ~0x10));
    }

    /**
     * Configure the FIFO (FIFO_CTRL_REG).
     * @param {number} mode      FIFO mode 0=Bypass, 1=FIFO, 2=Stream, 3=Bypass-to-Stream, 7=Stream-to-FIFO.
     * @param {number} watermark Watermark threshold 0-31 (WTM[4:0]).
     */
    async configureFifo(mode, watermark) {
        const validModes = [0, 1, 2, 3, 7];
        if (!validModes.includes(mode)) throw new Error('mode must be 0, 1, 2, 3, or 7');
        if (watermark < 0 || watermark > 31) throw new Error('watermark must be 0..31');
        const ctrl5 = (await this._readReg(_REG_CTRL_REG5, 1))[0] | 0x40;
        await this._writeReg(_REG_CTRL_REG5, ctrl5);
        await this._writeReg(_REG_FIFO_CTRL, ((mode & 0x7) << 5) | (watermark & 0x1F));
    }

    /**
     * Enable or disable the FIFO (FIFO_EN bit in CTRL_REG5).
     * @param {boolean} enable True to enable FIFO, False to disable and clear to bypass.
     */
    async enableFifo(enable) {
        const ctrl5 = (await this._readReg(_REG_CTRL_REG5, 1))[0];
        if (enable) {
            await this._writeReg(_REG_CTRL_REG5, ctrl5 | 0x40);
        } else {
            await this._writeReg(_REG_CTRL_REG5, ctrl5 & ~0x40);
            await this._writeReg(_REG_FIFO_CTRL, 0x00);
        }
    }

    /**
     * Read number of unread samples in FIFO (FIFO_SRC_REG FSS[4:0]).
     * @returns {Promise<number>} Number of stored samples (0-31).
     */
    async fifoLevel() {
        return (await this._readReg(_REG_FIFO_SRC, 1))[0] & 0x1F;
    }

    /**
     * Read all available FIFO samples and return as rad/s tuples.
     * @returns {Promise<Array<[number, number, number]>>} List of (x_rad_s, y_rad_s, z_rad_s).
     */
    async readFifo() {
        const n = await this.fifoLevel();
        if (n === 0) return [];
        const sens = this._sensitivity();
        const buf = await this._readReg(_REG_OUT_X_L, n * 6);
        const out = [];
        for (let i = 0; i < n; i++) {
            const o = i * 6;
            const x = _int16Le(buf, o) * sens * _DPS_TO_RAD;
            const y = _int16Le(buf, o + 2) * sens * _DPS_TO_RAD;
            const z = _int16Le(buf, o + 4) * sens * _DPS_TO_RAD;
            out.push([x, y, z]);
        }
        return out;
    }

    /**
     * Set the power mode (CTRL_REG1 PD and axis enable bits).
     * @param {string} mode 'normal' (PD=1, all axes on), 'sleep' (PD=1, all axes off),
     *                       or 'power_down' (PD=0).
     */
    async setPowerMode(mode) {
        let ctrl1 = (await this._readReg(_REG_CTRL_REG1, 1))[0];
        if (mode === L3GD20HFull.POWER_NORMAL) {
            ctrl1 = (ctrl1 & 0xF0) | 0x0F;
        } else if (mode === L3GD20HFull.POWER_SLEEP) {
            ctrl1 = (ctrl1 & 0xF8) | 0x08;
        } else if (mode === L3GD20HFull.POWER_POWERDOWN) {
            ctrl1 &= 0xF7;
        } else {
            throw new Error("mode must be 'normal', 'sleep', or 'power_down'");
        }
        await this._writeReg(_REG_CTRL_REG1, ctrl1);
    }
}

module.exports = { L3GD20HMinimal, L3GD20HFull };

File: cpp/src/chips/gyroscope/L3gd20h.cpp
#include "L3gd20h.h"
#include <math.h>
#include <string.h>

#ifdef ARDUINO
#include <Arduino.h>
static inline void delay_ms(unsigned long ms) { delay(ms); }
#elif defined(__ZEPHYR__)
#include <zephyr/kernel.h>
static inline void delay_ms(unsigned long ms) { k_sleep(K_MSEC(ms)); }
#elif defined(ESP_PLATFORM)
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>
static inline void delay_ms(unsigned long ms) { vTaskDelay(pdMS_TO_TICKS(ms)); }
#elif __has_include(<pico/time.h>)
#include <pico/time.h>
static inline void delay_ms(unsigned long ms) { sleep_ms(ms); }
#else
#include <unistd.h>
static inline void delay_ms(unsigned long ms) { usleep(ms * 1000UL); }
#endif

static constexpr float kPi = 3.141592653589793f;

L3gd20hMinimal::L3gd20hMinimal(Connection& connection, bool spi)
    : _connection(connection), _spi(spi), _full_scale(250) {
    uint8_t who = 0;
    _read_reg(REG_WHO_AM_I, &who, 1);
    if (who != WHO_AM_I_L3GD20 && who != WHO_AM_I_L3GD20H) {
        return;
    }
    _write_reg(REG_CTRL_REG4, CTRL_REG4_DEFAULT);
    _write_reg(REG_CTRL_REG1, CTRL_REG1_DEFAULT);
    delay_ms(250);
}

void L3gd20hMinimal::_write_reg(uint8_t reg, uint8_t value) {
    uint8_t addr = _spi ? (reg & 0x3F) : reg;
    uint8_t buf[2] = { addr, value };
    _connection.write(buf, 2);
}

void L3gd20hMinimal::_read_reg(uint8_t reg, uint8_t* buf, uint8_t len) {
    if (_spi) {
        uint8_t addr = reg | 0xC0;
        _connection.write_read(&addr, 1, buf, len);
    } else if (len > 1) {
        uint8_t addr = reg | 0x80;
        _connection.write_read(&addr, 1, buf, len);
    } else {
        _connection.write_read(&reg, 1, buf, len);
    }
}

float L3gd20hMinimal::_sensitivity() const {
    switch (_full_scale) {
        case 250:  return 8.75e-3f;
        case 500:  return 17.5e-3f;
        case 2000: return 70.0e-3f;
        default:   return 8.75e-3f;
    }
}

int16_t L3gd20hMinimal::_int16_le(const uint8_t* data) {
    int16_t v = (int16_t)((uint16_t)data[0] | ((uint16_t)data[1] << 8));
    return v;
}

void L3gd20hMinimal::gyro(float& x_rad_s, float& y_rad_s, float& z_rad_s) {
    uint8_t raw[6] = {0, 0, 0, 0, 0, 0};
    _read_reg(REG_OUT_X_L, raw, 6);
    float sens = _sensitivity();
    float x_dps = _int16_le(&raw[0]) * sens;
    float y_dps = _int16_le(&raw[2]) * sens;
    float z_dps = _int16_le(&raw[4]) * sens;
    x_rad_s = x_dps * (kPi / 180.0f);
    y_rad_s = y_dps * (kPi / 180.0f);
    z_rad_s = z_dps * (kPi / 180.0f);
}

// L3gd20hFull

L3gd20hFull::L3gd20hFull(Connection& connection, bool spi)
    : L3gd20hMinimal(connection, spi), _odr(0), _bw(0), _threshold_raw(0) {
}

void L3gd20hFull::configure(uint8_t odr, uint8_t bw, uint8_t full_scale) {
    if (odr > 3) odr = 3;
    if (bw > 3) bw = 3;
    if (full_scale > 2) full_scale = 2;
    _odr = odr;
    _bw = bw;
    static const uint16_t fs_map[3] = { 250, 500, 2000 };
    _full_scale = fs_map[full_scale];
    uint8_t ctrl1 = CTRL_REG1_DEFAULT | ((_odr & 0x3) << 6) | ((_bw & 0x3) << 4);
    _write_reg(REG_CTRL_REG1, ctrl1);
    uint8_t ctrl4 = CTRL_REG4_DEFAULT | ((full_scale & 0x3) << 4);
    _write_reg(REG_CTRL_REG4, ctrl4);
}

void L3gd20hFull::gyro_raw(int16_t& x_raw, int16_t& y_raw, int16_t& z_raw) {
    uint8_t raw[6] = {0, 0, 0, 0, 0, 0};
    _read_reg(REG_OUT_X_L, raw, 6);
    x_raw = _int16_le(&raw[0]);
    y_raw = _int16_le(&raw[2]);
    z_raw = _int16_le(&raw[4]);
}

int8_t L3gd20hFull::temperature() {
    uint8_t v = 0;
    _read_reg(REG_OUT_TEMP, &v, 1);
    return (int8_t)v;
}

bool L3gd20hFull::data_ready() {
    uint8_t v = 0;
    _read_reg(REG_STATUS, &v, 1);
    return (v & 0x08) != 0;
}

void L3gd20hFull::configure_hp_filter(uint8_t mode, uint8_t cutoff) {
    if (mode > 3) mode = 3;
    if (cutoff > 15) cutoff = 15;
    uint8_t ctrl2 = ((mode & 0x3) << 4) | (cutoff & 0x0F);
    _write_reg(REG_CTRL_REG2, ctrl2);
}

void L3gd20hFull::enable_hp_filter(bool enable) {
    uint8_t ctrl5 = 0;
    _read_reg(REG_CTRL_REG5, &ctrl5, 1);
    if (enable) {
        ctrl5 |= 0x10;
    } else {
        ctrl5 &= ~0x10;
    }
    _write_reg(REG_CTRL_REG5, ctrl5);
}

void L3gd20hFull::configure_fifo(uint8_t mode, uint8_t watermark) {
    static const uint8_t valid_modes[] = { 0, 1, 2, 3, 7 };
    bool valid = false;
    for (size_t i = 0; i < sizeof(valid_modes); i++) {
        if (mode == valid_modes[i]) { valid = true; break; }
    }
    if (!valid) mode = 0;
    if (watermark > 31) watermark = 31;
    uint8_t ctrl5 = 0;
    _read_reg(REG_CTRL_REG5, &ctrl5, 1);
    ctrl5 |= 0x40;
    _write_reg(REG_CTRL_REG5, ctrl5);
    uint8_t fifo_ctrl = ((mode & 0x7) << 5) | (watermark & 0x1F);
    _write_reg(REG_FIFO_CTRL, fifo_ctrl);
}

void L3gd20hFull::enable_fifo(bool enable) {
    uint8_t ctrl5 = 0;
    _read_reg(REG_CTRL_REG5, &ctrl5, 1);
    if (enable) {
        ctrl5 |= 0x40;
    } else {
        ctrl5 &= ~0x40;
        _write_reg(REG_FIFO_CTRL, 0x00);
    }
    _write_reg(REG_CTRL_REG5, ctrl5);
}

uint8_t L3gd20hFull::fifo_level() {
    uint8_t v = 0;
    _read_reg(REG_FIFO_SRC, &v, 1);
    return v & 0x1F;
}

uint8_t L3gd20hFull::read_fifo(float* out_x, float* out_y, float* out_z, uint8_t max_samples) {
    uint8_t n = fifo_level();
    if (n == 0 || max_samples == 0) return 0;
    if (n > max_samples) n = max_samples;
    float sens = _sensitivity();
    uint8_t* data = new uint8_t[n * 6];
    _read_reg(REG_OUT_X_L, data, n * 6);
    for (uint8_t i = 0; i < n; i++) {
        uint8_t offset = i * 6;
        float x_dps = _int16_le(&data[offset]) * sens;
        float y_dps = _int16_le(&data[offset + 2]) * sens;
        float z_dps = _int16_le(&data[offset + 4]) * sens;
        out_x[i] = x_dps * (kPi / 180.0f);
        out_y[i] = y_dps * (kPi / 180.0f);
        out_z[i] = z_dps * (kPi / 180.0f);
    }
    delete[] data;
    return n;
}

void L3gd20hFull::set_power_mode(const char* mode) {
    if (!mode) return;
    if (strcmp(mode, POWER_NORMAL) == 0) {
        uint8_t ctrl1 = 0;
        _read_reg(REG_CTRL_REG1, &ctrl1, 1);
        ctrl1 = (ctrl1 & 0xF0) | 0x0F;
        _write_reg(REG_CTRL_REG1, ctrl1);
    } else if (strcmp(mode, POWER_SLEEP) == 0) {
        uint8_t ctrl1 = 0;
        _read_reg(REG_CTRL_REG1, &ctrl1, 1);
        ctrl1 = (ctrl1 & 0xF8) | 0x08;
        _write_reg(REG_CTRL_REG1, ctrl1);
    } else if (strcmp(mode, POWER_POWERDOWN) == 0) {
        uint8_t ctrl1 = 0;
        _read_reg(REG_CTRL_REG1, &ctrl1, 1);
        ctrl1 &= 0xF7;
        _write_reg(REG_CTRL_REG1, ctrl1);
    }
}